core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1244,6 +1244,61 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When either the inner or outer stream fails, the entire streams fails.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+      Stream.eval(Deferred[F2, Unit]).flatMap { haltOuter =>
+        Stream.eval(async.unboundedQueue[F2, Option[Either[Throwable, Chunk[O2]]]]).flatMap {
+          queue =>
+            Stream.eval(async.unboundedQueue[F2, Option[(O, Deferred[F2, Unit])]]).flatMap {
+              pulse =>
+                Stream.eval(Semaphore(1)).flatMap { guard =>
+                  queue.dequeue.unNoneTerminate.rethrow
+                    .flatMap(Stream.chunk(_))
+                    .concurrently(pulse.dequeue.unNoneTerminate
+                      .onFinalize(queue.enqueue1(None))
+                      .evalMap {
+                        case (o, haltInner) =>
+                          guard.withPermit(
+                            f(o).chunks
+                              .evalMap(x => queue.enqueue1(Right(x).some))
+                              .interruptWhen(haltInner.get.attempt)
+                              .interruptWhen(haltOuter.get.attempt)
```

Comment:
I'm not sure `.interruptWhen(haltOuter.get.attempt` is necessary. 
As far as I can see it's only completed in `handleError` of this same stream, but if there's an error this stream will already be terminated